src/main/kotlin/org/rust/cargo/toolchain/RsToolchain.kt

```diff
@@ -5,42 +5,77 @@
 import com.intellij.util.io.exists
+import com.intellij.util.net.HttpConfigurable
 import com.intellij.util.text.SemVer
+import org.rust.cargo.CargoConstants
 import org.rust.cargo.toolchain.flavors.RsToolchainFlavor
 import org.rust.cargo.toolchain.tools.Cargo
-import org.rust.cargo.util.hasExecutable
-import org.rust.cargo.util.pathToExecutable
-import org.rust.stdext.isExecutable
+import org.rust.openapiext.GeneralCommandLine
+import org.rust.openapiext.withWorkDirectory
 import java.io.File
 import java.nio.file.Path
+import java.nio.file.Paths
 
-open class RsToolchain(val location: Path) {
-    val presentableLocation: String = pathToExecutable(Cargo.NAME).toString()
+interface RsToolchainProvider {
+    fun isApplicable(homePath: Path): Boolean
+    fun getToolchain(homePath: Path): RsToolchain?
+
+    companion object {
+        private val EP_NAME: ExtensionPointName<RsToolchainProvider> =
+            ExtensionPointName.create("org.rust.toolchainProvider")
+
+        fun getToolchain(homePath: Path): RsToolchain? =
+            EP_NAME.extensionList.find { it.isApplicable(homePath) }?.getToolchain(homePath)
+    }
+}
+
+abstract class RsToolchain(val location: Path) {
+    val presentableLocation: String get() = pathToExecutable(Cargo.NAME).toString()
+
+    abstract val fileSeparator: String
+
+    abstract val executionTimeoutInMilliseconds: Int
 
     fun looksLikeValidToolchain(): Boolean = RsToolchainFlavor.getFlavor(location) != null
 
+    /**
+     * Patches passed command line to make it runnable in remote context.
+     */
+    abstract fun <T : GeneralCommandLine> patchCommandLine(commandLine: T): T
+
+    abstract fun startProcess(commandLine: GeneralCommandLine): ProcessHandler
+
+    abstract fun toLocalPath(remotePath: String): String
+
+    abstract fun expandUserHome(remotePath: String): String
+
+    protected abstract fun getExecutableName(toolName: String): String
+
     // for executables from toolchain
-    fun pathToExecutable(toolName: String): Path = location.pathToExecutable(toolName)
+    abstract fun pathToExecutable(toolName: String): Path
 
     // for executables installed using `cargo install`
     fun pathToCargoExecutable(toolName: String): Path {
         // Binaries installed by `cargo install` (e.g. Grcov, Evcxr) are placed in ~/.cargo/bin by default:
         // https://doc.rust-lang.org/cargo/commands/cargo-install.html
         // But toolchain root may be different (e.g. on Arch Linux it is usually /usr/bin)
-        val path = pathToExecutable(toolName)
-        if (path.exists()) return path
-
-        val exeName = if (SystemInfo.isWindows) "$toolName.exe" else toolName
-        val cargoBinPath = File(FileUtil.expandUserHome("~/.cargo/bin")).toPath()
-        return cargoBinPath.resolve(exeName).toAbsolutePath()
+        val exePath = pathToExecutable(toolName)
+        if (exePath.exists()) return exePath
+        val cargoBin = expandUserHome("~/.cargo/bin")
```

Comment:
I suppose it may cause some freeze with wsl toolchain here because `expandUserHome` runs an external process to get value of `HOME` variable.
I don't insist to fix it in this PR but it would be great to collect such issues, check them and fix if possible. For example here, we can cache value of `HOME` variable (or not?)

Reply:
Now I use the lazy `WSLDistribution.getUserHome` method inside of `expandUserHome`.